feat(type-checking): Add type-checking pre-commit hooks

alveifbklsiu259 opened the pull request:
feat(type-checking): Add type-checking pre-commit hooks

### SUMMARY

This PR introduces some pre-commit hooks to check TypeScript files using `tscw-config`. This enhancement aims to streamline the development process by running type checks **only on changed files**.

#### NOTES

- Since different packages (e.g., `superset-frontend`, `superset-websocket`...) use their own `tsconfig.json` and may require different declaration files, ~~multiple type-checking hooks are used~~. (edit: after discussion, only one hook (`type-check-frontend`) is used, if future requirement changes, one can always reference [commits](https://github.com/apache/superset/pull/32261/commits)).
- JS/JSX files will be processed by the type-checking hooks, but `checkJs` is not enabled by default. To enable type-checking for JS/JSX files, either include `//@ts-check` at the top of each JS/JSX file or enable it in the package's `tsconfig.json`.
- Each package, such as `superset-frontend`, `superset-websocket`... will include all files ending with `.d.ts` in their respective directories (recursively). However, if one needs to include third-party declaration files, one can either import them directly in the files that need them or declare them in the `type` field of the package's `tsconfig.json`. For example `import "@testing-library/jest-dom"`
- Under the hood, what this hook does it create a `tsconfig.json` from the current package (`superset-frontend/tsconfig.json` for `superset-frontend`), remove the `include` filed, and include all the files passed to this hook (and some declaration files) in the `files` field. This means that ANY files passed to the hook will be type-cheked even if they are not included in the original `tsconfig.json`'s `include` field. For example, at the time of writing, `superset-frontend/tsconfig.json` does not include files from `./*/storybook/**/*`, but if one modifies `superset-frontend/packages/superset-ui-demo/storybook/stories/superset-ui-color/RenderPalettes.tsx`, this file will still be type-checked. IMHO type-checking the files even though they will not be compiled to javascript is not a bad idea, but if one really wants to disable this behavior, one can:
   - Use `// @ts-nocheck`.
   - Prevent those files from being passed to the hook, for example, use `exclude: ^(superset-frontend/cypress-base/|superset-frontend/.*/storybook/)` in `pre-commit-config.yaml`.



### DISCLOSURE

I am the author of `tscw-config`. I genuinely believe that this library offers an effective solution to the issue discussed in [Issue #27379](https://github.com/microsoft/TypeScript/issues/27379#issuecomment-2565940933).

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
<!--- Skip this if not applicable -->

### TESTING INSTRUCTIONS
<!--- Required! What steps can be taken to manually verify the changes? -->

### ADDITIONAL INFORMATION
<!--- Check any relevant boxes with "x" -->
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [x] Introduces new feature or API
- [ ] Removes existing feature or API

alveifbklsiu259 committed:
feat(type-checking): Add type-checking pre-commit hook for frontend

Added the script as a pre-commit hook to ensure type-checking before commits.

alveifbklsiu259 committed:
feat(type-checking): Add type-checking pre-commit hook for websocket

alveifbklsiu259 committed:
feat(type-checking): Add type-checking pre-commit hook for embedded-sdk

alveifbklsiu259 committed:
feat(type-checking): Add type-checking pre-commit hook for frontend/cypress

alveifbklsiu259 committed:
refactor(type-checking): correct log levels, improve performance, and add error context

- Updated log levels for stdout and stderr to reflect appropriate severity.
- Refactored fs operations to use asynchronous method.
- Enhanced error handling by adding context information.

sadpandajoe commented:
@mistercrunch don't we already have some sort of type checking? Adding you since you did a lot of work to clean up our pre-commits

mistercrunch commented on .pre-commit-config.yaml:
```diff
@@ -68,6 +68,39 @@ repos:
         language: system
         pass_filenames: true
         files: \.(js|jsx|ts|tsx)$
+  - repo: local
```
wondering if you could get away with a single entry here (?)

mistercrunch commented:
@sadpandajoe yes we do in CI, but this add the checks to `pre-commit` so issues can be caught at commit time, which is better. This adds in a fair amount of complexity to deliver on it. 

Wondering if it would make sense to only do it on superset-frontend/**/js-and-ts-files to keep pre-commit.yml a bit simpler.

@alveifbklsiu259  is it all super fast?

mistercrunch commented on .pre-commit-config.yaml:
```diff
@@ -68,6 +68,39 @@ repos:
+    hooks:
+      - id: type-checking-frontend
+        name: Type-Checking (Frontend)
+        entry: ./scripts/check-type.js
```
From my understanding this will check all the files, not just the ones touched in the commit is that right?  Same for all other entries. 

If so it seems this will greatly slow down commit speed.

You can see how I did it for eslint, but by using `$@`, I'm able to pass the list of files to check to the command, so it will only check the files touched in that specific commit.

mistercrunch replied:
know that `npm run type` take 10-30 seconds, we don't want to add this kind of lag for each commit, especially when you're not even touching these files. 

Key here is to somehow pass the list of files touched in the commit (provided by pre-commit), and only run the checks on those files. Now guessing `tsc` may have to go and import all the related files referenced in that seed file, and could end up taking quite some time if you end up covering files-that-import-files-that-import-other-files, ...

alveifbklsiu259 replied:
> From my understanding this will check all the files, not just the ones touched in the commit is that right?

It seems that we don't have to explicitly pass `$@`, pre-commit will only passed the touched files:

```js
// check-type.js
void (async () => {
  const args = process.argv.slice(2);
  console.log(args)
  exit(1)
```
![1](https://github.com/user-attachments/assets/cd3037b8-9f3f-4ddb-aac2-0e9fbb091d2d)

> is it all super fast?

I wouldn't say it is **super fast**, but it is faster than checking all files.

![2](https://github.com/user-attachments/assets/d7951fa8-38f5-427a-972a-1b0cc09e9187)
![3](https://github.com/user-attachments/assets/c189919b-eb43-48c0-9c43-03bdb4a97d63)

> Key here is to somehow pass the list of files touched in the commit (provided by pre-commit), and only run the checks on those files.

This is exactly what this pre-commit hook does.

> ... may have to go and import all the related files referenced in that seed file, and could end up taking quite some time if you end up covering files-that-import-files-that-import-other-files

This is still true for this hook. For example:

```ts
// superset-frontend/src/components/AsyncAceEditor/index.tsx
import AsyncEsmComponent, {
  PlaceholderProps,
} from 'src/components/AsyncEsmComponent';
```

If `superset-frontend/src/components/AsyncAceEditor/index.tsx` is the only file in the staging area, running this hook will also check `src/components/AsyncEsmComponent/index.tsx` and the files it imports and so on.

> Wondering if it would make sense to only do it on superset-frontend/**/js-and-ts-files to keep pre-commit.yml a bit simpler.

Sure, as long as we only want to type-check the files under `superset-frontend`, but we still need to exclude `superset-frontend/cypress-base/`, since `superset-frontend/cypress-base/` uses a different `tsconfig.json` and it contains a conflicting declaration file. i.e. `superset-frontend/cypress-base/cypress/support/index.d.ts`

mistercrunch replied:
Oh nice. Great to hear it's all working. And I understand now that we need one entry per tsconfig.

alveifbklsiu259 committed:
fix(type-checking): Fix regular expression injection vulnerability

alveifbklsiu259 committed:
chore(type-checking): remove unnecessary pre-commit hooks

- Removed type-checking-websocket hook
- Removed type-checking-embedded-sdk hook
- Removed type-checking-cypress hook

alveifbklsiu259 commented on scripts/check-type.js:
```diff
@@ -0,0 +1,180 @@
+  }
+
+  const packageRootDir = getPackage(packageArg);
+  const packagePathRegex = new RegExp(`^${packageRootDir}\/`);
```
[fixed](https://github.com/apache/superset/pull/32261/commits/503378e5bf4e8b0ae54044e5699588756041a7d1).

alveifbklsiu259 committed:
chore(ci): disable type-checking-frontend pre-commit hook execution in CI pipeline

Disable it as it requires npm ci, plus type-check is handled by other job.

alveifbklsiu259 commented:
/korbit-review

alveifbklsiu259 committed:
fix(type-checking): improve error handling and process outputs

- Add check for module installation and provide clear error message if missing
- Ensure both stdout and stderr are logged if they contain content
- Add check for existence of tsconfig.json and provide clear error message if missing

alveifbklsiu259 commented:
/korbit-review

alveifbklsiu259 committed:
fix(type-checking): Enhance functions for error handling and performance

- getFilesRecursively: use `promise.all`, handle error.
- getPackage: handle error.

alveifbklsiu259 commented:
/korbit-review

alveifbklsiu259 committed:
refactor(type-checking): update exclusion logic for more fine-grained controll

alveifbklsiu259 commented:
/korbit-review

alveifbklsiu259 committed:
fix(type-checking): fix exclusion logic false positives

alveifbklsiu259 commented on scripts/check-type.js:
```diff
@@ -0,0 +1,244 @@
+      }
+    }
+
+    const recursiveResults = await Promise.all(recursivePromises);
```
@korbit-ai

After analyzing the memory impact, I found that the increase is not significant enough to warrant the added complexity of introducing a semaphore or concurrency limit. Implementing such a solution would complicate the codebase.

I believe the current implementation is sufficient for our use case and does not pose any immediate performance concerns. Maybe we will consider revisiting this approach if we encounter any memory-related issues in the future.

```ts
  const initial = process.memoryUsage();

  let declarationFiles = await getFilesRecursively(
    packageRootDir,
    DECLARATION_FILE_REGEX,
    excludedDeclarationDirs
  );
  const final = process.memoryUsage();

  const memoryDiff = {
    rss: `${Math.round((final.rss - initial.rss) / (1024 * 1024))} MB`,
    heapTotal: `${Math.round(
      (final.heapTotal - initial.heapTotal) / (1024 * 1024)
    )} MB`,
    heapUsed: `${Math.round(
      (final.heapUsed - initial.heapUsed) / (1024 * 1024)
    )} MB`,
    external: `${Math.round(
      (final.external - initial.external) / (1024 * 1024)
    )} MB`,
  };

  console.log("Memory Difference:", memoryDiff); 
  // Memory Difference: { rss: '16 MB', heapTotal: '9 MB', heapUsed: '3 MB', external: '0 MB' }
```

mistercrunch commented on scripts/check-type.js:
```diff
@@ -0,0 +1,220 @@
+
+  const packageRootDirAbsolute = join(SUPERSET_ROOT, packageRootDir);
+  const tsConfig = getTsConfig(packageRootDirAbsolute);
+  const command = `--noEmit --allowJs --composite false --project ${tsConfig} ${argsStr} ${declarationFilesStr}`;
```
Agreed, not an issue here

mistercrunch reviewed (approved):
Thanks for doing this! This will certainly save extra CI cycles by catching type errors at commit time.

rusackas reviewed (approved):
Looks safe from a repo secrets perspective. Thanks for the hard work here, and looking forward to the performance bump!
